client/src/unifyfs-sysio.c

```diff
@@ -636,45 +601,23 @@ int unifyfs_fd_write(int fd, off_t pos, const void* buf, size_t count)
     /* TODO: check that file is open for writing */
 
     /* get current file size before extending the file */
-    off_t filesize = unifyfs_fid_size(fid);
+    off_t filesize = unifyfs_fid_local_size(fid);
 
     /* compute new position based on storage type */
     off_t newpos;
     unifyfs_filemeta_t* meta = unifyfs_get_meta_from_fid(fid);
-    if (meta->storage == FILE_STORAGE_FIXED_CHUNK) {
-        /* extend file size and allocate chunks if needed */
-        newpos = pos + (off_t) count;
-        int extend_rc = unifyfs_fid_extend(fid, newpos);
-        if (extend_rc != UNIFYFS_SUCCESS) {
-            return extend_rc;
-        }
-
-        /* fill any new bytes between old size and pos with zero values */
-        if (filesize < pos) {
-            off_t gap_size = pos - filesize;
-            int zero_rc = unifyfs_fid_write_zero(fid, filesize, gap_size);
-            if (zero_rc != UNIFYFS_SUCCESS) {
-                return zero_rc;
-            }
-        }
-    } else if (meta->storage == FILE_STORAGE_LOGIO) {
-        newpos = filesize + (off_t)count;
-        int extend_rc = unifyfs_fid_extend(fid, newpos);
-        if (extend_rc != UNIFYFS_SUCCESS) {
-            return extend_rc;
-        }
-    } else {
-        return UNIFYFS_ERROR_IO;
+
+    newpos = filesize + (off_t)count - pos;
+    int extend_rc = unifyfs_fid_extend(fid, newpos);
+    if (extend_rc != UNIFYFS_SUCCESS) {
+        return extend_rc;
     }
 
     /* finally write specified data to file */
     int write_rc = unifyfs_fid_write(fid, pos, buf, count);
     if (write_rc == 0) {
         meta->needs_sync = 1;
-        if (meta->storage == FILE_STORAGE_LOGIO) {
-            meta->size = newpos;
-            meta->log_size = pos + count;
-        }
+        meta->local_size = MAX(meta->local_size, pos + count);
     }
```

Comment:
The logic that was tracking log size on writes was obfuscated in the old code here.  For UnifyFS, the meta->size field was actually recording the size of the log, not the size of the file.  The unifyfs_fid_size(fid) line here is returning the current size of the log and storing that into "filesize".  So then filesize holds the size of the log, not the file.

We then drop into the "else if storage==LOGIO" branch and compute the size that the log will be after we append the data associated with the current write operation, which we store in newpos ```newpos = filesize + count```.  The newpos value is then passed to unifyfs_fid_extend.  That function determines whether the current set of chunks assigned to the file are sufficient to hold the new data that we're about to append, and if not, it allocates more chunks to the log as needed.

Finally data is appended to the log in fid_write and then the meta->size field is updated to record the new size of the log ```meta->size = newpos```.

The meta->log_size variable was never fully implemented as a replacement, so it was essentially a NOP.

Obviously, all of that logic is hard to make sense of due to the variable names.

Reply:
Had the old code use better names, this would have been something like:
```
logsize = unifyfs_fid_logsize(fid) // returns meta->log_size
newlogsize = logsize + count // size of log after appending new data
unifyfs_fid_extend(fid, newlogsize) // allocate new chunks for log, if needed
unifyfs_fid_write(fid, pos, but, count) // copy data into log
meta->log_size = newlogsize; // update size of log
```

Reply:
Ok, I think I get it.  Log_size is basically the sum of all 'counts' when we do a write(..., count).  It makes sense that that calculation would be used for allocating more chunks.  I'll add `log_size` back in and use it for the allocation.

Reply:
@adammoody thanks for catching this.